c2rust-analyze/src/util.rs

```diff
@@ -203,7 +205,7 @@ fn builtin_callee(tcx: TyCtxt, did: DefId) -> Option<Callee> {
             if tcx.impl_trait_ref(parent_did).is_some() {
                 return None;
             }
-            let parent_impl_ty = tcx.type_of(parent_did);
+            let parent_impl_ty = EarlyBinder(tcx.type_of(parent_did)).subst(tcx, substs);
```

Comment:
> could you comment on why this is necessary
>
> _Originally posted by @aneksteind in https://github.com/immunant/c2rust/pull/876#discussion_r1190176434_

Previously, for a call like `(x: *mut i32).offset(...)`, we would look up the `offset` method, find its containing `impl<T> *mut T { ... }`, take the impl type `*mut T`, extract the pointee type `T`, and record that as the `pointee_ty` in `Callee::PtrOffset`.  But having `T` here is completely useless (all `PtrOffset`s had the same `pointee_ty` due to this bug).  What we actually want in this example is for `pointee_ty` to be `i32`, which is the pointee type for this particular `offset` call.  This substitution takes the impl type `*mut T` and substitutes in the `substs` used for this particular call, which in our example produces `*mut i32`.  From there, the remaining steps extract the desired `i32` pointee type, and we end up with `PtrOffset { pointee_ty: i32 }`.